dll/win32/shell32/folders/CMyDocsFolder.cpp

```diff
@@ -34,34 +34,87 @@ CMyDocsFolder::~CMyDocsFolder()
         SHFree(m_pidlInner);
 }
 
+HRESULT WINAPI CMyDocsFolder::FinalConstruct()
+{
+    m_pidlInner = _ILCreateMyDocuments();
+
+    if (!m_pidlInner)
+        return E_OUTOFMEMORY;
+
+    return S_OK;
+}
+
+HRESULT CMyDocsFolder::EnsureFolder()
+{
+    ATLASSERT(m_pidlInner);
+
+    if (m_pisfInner)
+        return S_OK;
+
+
+    HRESULT hr = SHELL32_CoCreateInitSF(m_pidlInner,
+                                        &CLSID_ShellFSFolder,
+                                        CSIDL_PERSONAL,
+                                        IID_PPV_ARG(IShellFolder2, &m_pisfInner));
+
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    return S_OK;
+}
+
 HRESULT WINAPI CMyDocsFolder::ParseDisplayName(HWND hwndOwner, LPBC pbc, LPOLESTR lpszDisplayName,
         ULONG *pchEaten, PIDLIST_RELATIVE *ppidl, ULONG *pdwAttributes)
 {
+    HRESULT hr = EnsureFolder();
```

Comment:
It seems really annoying to have to call this from every function. I recall looking at what Windows does and, at least for some folders, it seems to initialize the class with a valid pidl in the constructor. Could that be the right thing to do?

Reply:
That is basically what I do, but from that pidl we have to create a shell folder.
And this cannot be done from the constructor, because someone can also call Initialize.
Given, we could destroy the inner shell folder again, but that seems a huge waste of cycles.